contrib/babelfishpg_tsql/src/pl_funcs.c

```diff
@@ -864,8 +865,11 @@ pltsql_free_function_memory(PLtsql_function *func)
 	 * fn_extra pointers to it).
 	 */
 	if (func->fn_cxt)
-		MemoryContextDelete(func->fn_cxt);
```

Comment:
The variable func->fn_cxt is a pointer. MemoryContextDelete will frees up the memory pointed by func->fn_cxt and it doesn't free func->fn_cxt itself. You can't access func->fn_cxt if you do a pfree(func). But, that's not happening in this case.
```
func->fn_cxt = NULL
```
The above code just changes the pointer value to NULL. So, I don't see why it'll cause an usage of memory after free issue.

Reply:
I well could have misread the existing logic in `pl_comp.c`, my reasoning was based on the following understanding:
 1. [func_cxt is set as a current memory context](https://github.com/babelfish-for-postgresql/babelfish_extensions/blob/aaa7127d18f01188b8945029eb5a27dc2298f378/contrib/babelfishpg_tsql/src/pl_comp.c#L1055)
 2. [function structure itself is allocated in current memory contex](https://github.com/babelfish-for-postgresql/babelfish_extensions/blob/aaa7127d18f01188b8945029eb5a27dc2298f378/contrib/babelfishpg_tsql/src/pl_comp.c#L1058)
 3. [the same memory context is assigned to fn_cxt field of the function](https://github.com/babelfish-for-postgresql/babelfish_extensions/blob/aaa7127d18f01188b8945029eb5a27dc2298f378/contrib/babelfishpg_tsql/src/pl_comp.c#L1066)

I assumed that, after "MemoryContextDelete will free up the memory pointed by func->fn_cxt" and when the function struct itself was allocated in this memory (using the same context) - then writes to the fields of this function structure will be "use after free" situation.

Reply:
Actually, you are right. The function memory itself is being allocated in func->fn_cxt. And, we're accessing it from more than one place after it's freed. Apart from the one you've fixed, we also need to fix [this](https://github.com/babelfish-for-postgresql/babelfish_extensions/blob/aaa7127d18f01188b8945029eb5a27dc2298f378/contrib/babelfishpg_tsql/src/pl_comp.c#L224). Here, delete_function is freeing up the memory if usage count is 0, then again accessing the variable. Could you please include this in your fix as well?

Reply:
Sorry for delay, will be able to update the PR in a few days.

Reply:
Moved `use_count` check to happen before the `delete_function` call, the logic is kept the same.